bftclient/src/bft_client.cpp

```diff
@@ -12,28 +12,71 @@
 #include "bftclient/bft_client.h"
 #include "bftengine/ClientMsgs.hpp"
 #include "assertUtils.hpp"
+#include "secrets_manager_enc.h"
+#include "secrets_manager_plain.h"
+
+using namespace concord::diagnostics;
+using namespace concord::secretsmanager;
+using namespace bftEngine;
+using namespace bftEngine::impl;
 
 namespace bft::client {
 
-// This function creates a ClientRequestMsg or a ClientPreProcessRequestMsg depending upon config.
-//
-// Since both of these are just instances of a `ClientRequestMsgHeader` followed by the message
-// data, we construct them here, rather than relying on the type constructors embedded into the
-// bftEngine impl. This allows us to not have to link with the bftengine library, and also allows us
-// to return the messages as vectors with proper RAII based memory management.
-Msg makeClientMsg(const RequestConfig& config, Msg&& request, bool read_only, uint16_t client_id) {
+Client::Client(std::unique_ptr<bft::communication::ICommunication> comm, const ClientConfig& config)
+    : communication_(std::move(comm)),
+      config_(config),
+      quorum_converter_(config_.all_replicas, config_.f_val, config_.c_val),
+      expected_commit_time_ms_(config_.retry_timeout_config.initial_retry_timeout.count(),
+                               config_.retry_timeout_config.number_of_standard_deviations_to_tolerate,
+                               config_.retry_timeout_config.max_retry_timeout.count(),
+                               config_.retry_timeout_config.min_retry_timeout.count(),
+                               config_.retry_timeout_config.samples_per_evaluation,
+                               config_.retry_timeout_config.samples_until_reset,
+                               config_.retry_timeout_config.max_increasing_factor,
+                               config_.retry_timeout_config.max_decreasing_factor),
+      metrics_(config.id),
+      histograms_(std::unique_ptr<Recorders>(new Recorders(config.id))) {
+  // secrets_manager_config can be set only if transaction_signing_private_key_file_path is set
+  if (config.secrets_manager_config) ConcordAssert(config.transaction_signing_private_key_file_path != std::nullopt);
+  communication_->setReceiver(config_.id.val, &receiver_);
+  communication_->Start();
```

Comment:
does it matter if we start the communication before we finished initialization?

Reply:
@cloudnoize  - probably not, since the client won't send anything before ctor is done and the only initialization left is for the signer which belongs to sending part.
But think it would be better to move the start to the end in case of some code modifications that may change behavior.
thanks.